applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h

```diff
@@ -237,19 +233,18 @@ class GeometryUtilities
         KRATOS_ERROR_IF(r_edge_model_part.Nodes().size() != 0) << "ExtractEdgeNodes: The edge model part already has nodes!" << std::endl;
 
         for (auto& r_node_i : mrModelPart.Nodes()) {
-            Node<3>* p_node_i = &r_node_i;
-            GlobalPointersVector< Node<3 > >& r_node_i_neighbours = p_node_i->GetValue(NEIGHBOUR_NODES);
-            for( GlobalPointersVector< Node<3> >::iterator j =	r_node_i_neighbours.begin(); j != r_node_i_neighbours.end(); j++) {
-                GlobalPointersVector<Element>& r_element_neighbours = p_node_i->GetValue(NEIGHBOUR_ELEMENTS);
+            auto& r_node_i_neighbours = r_node_i.GetValue(NEIGHBOUR_NODES);
+            for(auto j = r_node_i_neighbours.begin(); j != r_node_i_neighbours.end(); j++) {
+                auto& r_element_neighbours = r_node_i.GetValue(NEIGHBOUR_ELEMENTS);
```

Comment:
```suggestion
                const auto& r_element_neighbours = r_node_i.GetValue(NEIGHBOUR_ELEMENTS);
```

Reply:
Compiler is warning an error with `const`  :(

Reply:
strange, this whole algorithm should be working with `const`, since only the `r_node_i` has to be non `const` since it is further used. All the other operations are just read operations.

Reply:
this works, not sure why the highlighted lines can not be made const, the pointervectors seem to have the const iterators iplemented. @philbucher  any idea?

```cpp
        for (auto& r_node_i : mrModelPart.Nodes()) {
            auto& r_node_i_neighbours = r_node_i.GetValue(NEIGHBOUR_NODES);  // does not work with const
            for(auto& r_node_j : r_node_i_neighbours) {
                auto& r_element_neighbours = r_node_i.GetValue(NEIGHBOUR_ELEMENTS);  // does not work with const
                int count = 0;
                for(const auto& r_elem_k : r_element_neighbours) {
                    const auto& r_element_geometry = r_elem_k.GetGeometry();
                    for(const auto& r_node_l : r_element_geometry) {
                        if (r_node_l.Id() == r_node_j.Id()) count ++;
                    }
                }
                if (count < 2){
                    r_edge_model_part.AddNode(&r_node_i);
                    break;
                }
            }
        }
```

Reply:
what are the error messages?

Reply:
also why can `r_node_i` not be `const`? Only its address is used to add it to the `r_edge_model_part`

Reply:
hm, i thought that in that case a "const" node is added to the new model part which should not be possible, but might be a misunderstanding from my side.

Reply:
> also why can `r_node_i` not be `const`? Only its address is used to add it to the `r_edge_model_part`

with `r_node_i` as `const` compiler throws the following error: :(

```
error: cannot convert ‘const Kratos::Node<3>*’ to ‘Kratos::Node<3>::Pointer’ {aka ‘Kratos::intrusive_ptr<Kratos::Node<3> >’}
```

Reply:
Another error message, using 
```
const auto& r_element_neighbours = r_node_i.GetValue(NEIGHBOUR_ELEMENTS);
```
in the suggested code snippet by @armingeiser 

```
/home/david/Software/Kratos/Kratos/applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:240:44:   required from here
/usr/include/boost/iterator/indirect_iterator.hpp:117:31: error: binding reference of type ‘boost::iterators::detail::iterator_facade_base<boost::iterators::indirect_iterator<__gnu_cxx::__normal_iterator<const Kratos::GlobalPointer<Kratos::Element>*, std::vector<Kratos::GlobalPointer<Kratos::Element>, std::allocator<Kratos::GlobalPointer<Kratos::Element> > > >, boost::use_default, boost::use_default, boost::use_default, boost::use_default>, Kratos::Element, boost::iterators::random_access_traversal_tag, Kratos::Element&, long int, true, true>::reference’ {aka ‘Kratos::Element&’} to ‘const Kratos::Element’ discards qualifiers
  117 |           return **this->base();
```

Reply:
While i expected that `r_node_i ` can not be made `const `, i still do not understand why its not possible to iterate over a `const` vector of  `NEIGHBOUR_ELEMENTS/NEIGHBOUR_NODES`.